torch_geometric/sampler/neighbor_sampler.py

```diff
@@ -396,106 +389,237 @@ def edge_sample(
     num_src_nodes: int,
     num_dst_nodes: int,
     disjoint: bool,
-    input_type: Optional[Tuple[str, str]] = None,
-    **kwargs,
+    input_type: Optional[Tuple[str, str, str]] = None,
+    node_time: Optional[Union[Tensor, Dict[str, Tensor]]] = None,
+    neg_sampling: Optional[NegativeSamplingConfig] = None,
 ) -> Union[SamplerOutput, HeteroSamplerOutput]:
     r"""Performs sampling from an edge sampler input, leveraging a sampling
     function of the same signature as `node_sample`."""
-    index, row, col, edge_label, edge_label_time = index
-    edge_label_index = torch.stack([row, col], dim=0)
-    negative_sampling_ratio = kwargs.get('negative_sampling_ratio', 0.0)
-
-    out = add_negative_samples(
-        edge_label_index,
-        edge_label,
-        edge_label_time,
-        num_src_nodes,
-        num_dst_nodes,
-        negative_sampling_ratio,
-    )
-    edge_label_index, edge_label, edge_label_time = out
-    num_seed_edges = edge_label_index.size(1)
-    seed_time = seed_time_dict = None
+    index, src, dst, edge_label, edge_label_time = index
+    src_time = dst_time = edge_label_time
+
+    assert edge_label_time is None or disjoint
+
+    num_pos = src.numel()
+    num_neg = 0
+
+    # Negative Sampling #######################################################
+
+    if neg_sampling is not None:
+        # When we are doing negative sampling, we append negative information
+        # of nodes/edges to `src`, `dst`, `src_time`, `dst_time`.
+        # Later on, we can easily reconstruct what belongs to positive and
+        # negative examples by slicing via `num_pos`.
+        num_neg = num_pos * neg_sampling.amount
+
+        if neg_sampling.is_binary():
+            # In the "binary" case, we randomly sample negative pairs of nodes.
+            if isinstance(node_time, dict):
+                src_node_time = node_time.get(input_type[0])
+            else:
+                src_node_time = node_time
+
+            src_neg = neg_sample(src, neg_sampling.amount, num_src_nodes,
+                                 src_time, src_node_time)
+            src = torch.cat([src, src_neg], dim=0)
+
+            if isinstance(node_time, dict):
+                dst_node_time = node_time.get(input_type[-1])
+            else:
+                dst_node_time = node_time
+
+            dst_neg = neg_sample(dst, neg_sampling.amount, num_dst_nodes,
+                                 dst_time, dst_node_time)
+            dst = torch.cat([dst, dst_neg], dim=0)
+
+            if edge_label is None:
+                edge_label = torch.ones(num_pos)
+            elif edge_label.min() == 0:
```

Comment:
Edge case:
`edge_label.min()==0` is `True` for the whole data, but for a given minibatch this isn't True